### PR TITLE
Fix integer division and encoding errors in tests under Py2 and improve pytest compatibility

### PR DESCRIPTION
**Pull request by oscarbenjamin**
#### Brief description of what is fixed or changed

This patch fixes a large number of fails seen under Python 2.7 with pytest.

These issues don't show up under `bin/test` because of the way it uses `compile/exec` rather than `__import__` for test modules. The changes are purely to test code. The changes are not noticeable in the tests when run under `bin/test` but are noticeable under `pytest` or when tests are imported and run normally under Python 2.7.

The reason these failures don't show up under `bin/test` is that the `bin/test` runner implicitly runs all tests with `from __future__ import division` due to the way that `compile/exec` works. It is possible to change this [https://www.python.org/dev/peps/pep-0264/](https://www.python.org/dev/peps/pep-0264/).

Most issues are to do with integer division e.g. using `1/3` rather than `S(1)/3`. This PR changes ~100 of those. In all cases checked I found only one where I thought that the intention really was for a floating point division. If these tests are imported and run directly with Python 2.7 then the errors can be seen:
```julia
In [1]: from sympy.algebras.tests.test_quaternion import test_quaternion_construction

In [2]: test_quaternion_construction()
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-2-6d040cc4afe9> in <module>()
----> 1 test_quaternion_construction()

/Users/enojb/current/sympy/sympy/sympy/algebras/tests/test_quaternion.py in test_quaternion_construction()
     15     q2 = Quaternion.from_axis_angle((sqrt(3)/3, sqrt(3)/3, sqrt(3)/3), 2*pi/3)
     16     assert q2 == Quaternion(Rational(1/2), Rational(1/2),
---> 17                             Rational(1/2), Rational(1,2))
     18 
     19     M = Matrix([[cos(x), -sin(x), 0], [sin(x), cos(x), 0], [0, 0, 1]])

AssertionError:
```

There is also one issue to do with encoding declaration. The test file `test_sympy_parser.py` contains non-ASCII characters but does not declare an encoding. Importing this directly under Python 2.7 shows the error:
```python
$ python
Python 2.7.10 (default, Oct  6 2017, 22:29:07) 
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sympy.parsing.tests.test_sympy_parser
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy/parsing/tests/test_sympy_parser.py", line 218
SyntaxError: Non-ASCII character '\xce' in file sympy/parsing/tests/test_sympy_parser.py on line 218, but no encoding declared; see http://python.org/dev/peps/pep-0263/ for details
```

#### Other comments

I have as part of finding these errors removed all `from __future__ import division` statements from the tests. I can put these back in but I don't think I should: removing them has exposed many bugs. I think it's best to leave them out and separately work out a way to ensure that `bin/test` doesn't conceal these bugs.

The reason these tests currently pass under `bin/test` is that `from __future__ import division` applies implicitly to all test modules. This means that e.g. `1/3` evaluates to a `float`. It just happens that the tests can pass with this `float` even though the test is intended to have an exact value `S(1)/3`. So even though the tests currently pass they are depending on floating point to do so even though nearly all cases should be exact.

#### Release Notes

* utilities
     * Fix unintentional int division across the test suite

### CODE REVIEW

**Comment by sidhantnagpal** [comment]
I think before merging this, the tests should be made to fail to detect all instances of it.

There are more such issues, an example being https://github.com/sympy/sympy/blob/master/sympy/geometry/tests/test_curve.py#L80.
The complete list of all such files (using division import from `__future__`) can be seen using
`git grep '__future__.*division' | grep 'tests/'`. It would probably be better to keep the tests file free from these imports.

**Comment by oscarbenjamin** [comment]
This issue is *not* caused by test files that have `from __future__ import division`. Somehow test files are having the future division injected into them by `bin/test`. I detected each of the errors above by running the tests under `pytest` which doesn't do this injection.

If we removing the `division` imports then it will probably show up more cases where `S()` is needed.

**Comment by sidhantnagpal** [comment]
The _last comment_ was in reference to the possibility of more cases of similar usage being present.
Inability to detect these, is a related issue, which is why I said that the tests could be made to fail to detect all instances (by fixing the test script).

**Comment by oscarbenjamin** [comment]
Ok I see what you mean. I've remove the `__future__` imports and I get 44 fails. I'll work through them and update here.

**Comment by asmeurer** [comment]
+1. Maybe we should have a test that ensures that every file has `from __future__ import print_function, division`.

**Comment by asmeurer** [comment]
The [astsearch](https://github.com/takluyver/astsearch) tool can be useful here, for instance, to find int/int.

**Comment by asmeurer** [comment]
Actually I guess astsearch can't wildcard based on type. But a simple grep should work.

File: sympy/parsing/tests/test_sympy_parser.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
+# -*- coding: utf-8 -*-
```
Comment (sidhantnagpal) [inline sympy/parsing/tests/test_sympy_parser.py]: Does pytest complain about the encoding as well?

Comment (oscarbenjamin) [reply sympy/parsing/tests/test_sympy_parser.py]: It's not pytest complaining. It's Python 2.7:

```console
$ python
Python 2.7.10 (default, Oct  6 2017, 22:29:07) 
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sympy.parsing.tests.test_sympy_parser
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy/parsing/tests/test_sympy_parser.py", line 218
SyntaxError: Non-ASCII character '\xce' in file sympy/parsing/tests/test_sympy_parser.py on line 218, but no encoding declared; see http://python.org/dev/peps/pep-0263/ for details
```

I'm not sure why this error doesn't show up when using `bin/test`

**Comment by oscarbenjamin** [comment]
Adding `from __future__ import print_function` isn't really helpful. The vast majority of the cases that show up as errors without it are mistakes where someone meant to write `S(1)/2` but instead wrote `1/2`. Adding the future import means that we get a float `0.5` on all Python versions but that isn't what is wanted on any Python versions. If the intention really is to have a float then the test should be written with a float literal so that that's clear. Otherwise it should be fixed with `S`.

**Comment by oscarbenjamin** [comment]
I get a failure in `test_ode` with
```console
$ pytest sympy/solvers/tests/test_ode.py -k test_linear_2eq_order2
```
The failure only shows up for this test with Python 2 under pytest. The test passes when imported normally and under `bin/test`. That one might turn out to be a bug on `pytest`.

I can't tell if there is a mistake in [here though](https://github.com/sympy/sympy/blob/sympy-1.3/sympy/solvers/tests/test_ode.py#L272).

**Comment by oscarbenjamin** [comment]
I've just pushed a new commit that removes all __future__ division imports and fixes all of the resulting test failures.

**Comment by oscarbenjamin** [comment]
I think I've fixed everything if the tests pass now.

**Comment by sidhantnagpal** [comment]
The number of files affected has grown big, but the changes look fine to me. The tests passed on restarting the affected job due to #15502.

It would be interesting to see why these failures (both for encoding as well as floats) didn't occur in the first place on Azure/Travis.

**Comment by oscarbenjamin** [comment]
I think it's because bin/test uses compile without specifying the flags. I'm not at a proper computer to check though:
https://www.python.org/dev/peps/pep-0264/

**Comment by oscarbenjamin** [comment]
I've just edited the PR description and title to explain more clearly what this patch is.

**Comment by sidhantnagpal** [comment]
+1. #15522 shows that tests don't fail, which can be fixed in a separate PR.

**Pull request merged**